LE Secure Connections Nordic

pan- opened the pull request.

pan- commented:
@paul-szczepanek-arm Could you review this PR ?

paul-szczepanek-arm commented on features/FEATURE_BLE/targets/TARGET_NORDIC/TARGET_NRF5/source/nRF5xPalSecurityManager.cpp:
```diff
@@ -727,9 +792,32 @@ bool nRF5xSecurityManager::sm_handler(const ble_evt_t *evt)
+
+            size_t key_size = public_key_coord_t::size();
+            public_key_coord_t peer_X(dhkey_request.p_pk_peer->pk, key_size);
+            public_key_coord_t peer_Y(dhkey_request.p_pk_peer->pk + key_size, key_size);
```
you don't need to pass in the size to the constructor and what's wrong with sizeof?

paul-szczepanek-arm commented on features/FEATURE_BLE/targets/TARGET_NORDIC/TARGET_NRF5/source/nRF5xPalSecurityManager.cpp:
```diff
@@ -727,9 +792,32 @@ bool nRF5xSecurityManager::sm_handler(const ble_evt_t *evt)
+            public_key_coord_t peer_X(dhkey_request.p_pk_peer->pk, key_size);
+            public_key_coord_t peer_Y(dhkey_request.p_pk_peer->pk + key_size, key_size);
+            public_key_coord_t sh_secret;
+            ble_gap_lesc_dhkey_t shared_secret;
```
the naming seems confusing, isn't sh_secret short for shared_secret?

pan- replied:
Both of these variables are used to hold the same shared secret. Use of an ArrayView would solve the problem and reduce stack usage.

paul-szczepanek-arm commented on features/FEATURE_BLE/targets/TARGET_NORDIC/TARGET_NRF5/source/nRF5xPalSecurityManager.cpp:
```diff
@@ -568,25 +630,26 @@ ble_error_t nRF5xSecurityManager::send_keypress_notification(
-    // FIXME: store locally
-    return BLE_ERROR_NOT_IMPLEMENTED;
-}
+    uint32_t err = sd_ble_gap_lesc_oob_data_get(
```
Didn't you say this was really slow? doesn't this block?

pan- replied:
That's not the slower part; computing the shared secret is. 
This does block application code but we have no choice outside spawning a worker thread which is not acceptable at the current minute.

pan- commented:
@paul-szczepanek-arm I've updated the PR; thanks for your comments.

paul-szczepanek-arm commented on features/FEATURE_BLE/targets/TARGET_NORDIC/TARGET_NRF5/source/nRF5xPalSecurityManager.cpp:
```diff
@@ -162,6 +179,7 @@ ble_error_t nRF5xSecurityManager::send_pairing_request(
     KeyDistribution initiator_dist,
     KeyDistribution responder_dist
 )  {
+    printf("nRF5xSecurityManager::send_pairing_request\r\n");
```
leftover debug

paul-szczepanek-arm commented on features/FEATURE_BLE/targets/TARGET_NORDIC/TARGET_NRF5/source/nRF5xCrypto.h:
```diff
@@ -0,0 +1,72 @@
+    LescCrypto();
+
+    ~LescCrypto();
+
```
missing doxygen

paul-szczepanek-arm commented on features/FEATURE_BLE/targets/TARGET_NORDIC/TARGET_NRF5/source/nRF5XCrypto.cpp:
```diff
@@ -0,0 +1,146 @@
+    const ArrayView<const uint8_t>& own_secret,
+    ArrayView<uint8_t> shared_secret
+) {
+    MBED_ASSERT(peer_X.size() == public_key_coord_t::size());
```
since there are no longer types to enforce size requirements like this need to be documented in doxygen

pan- commented on features/FEATURE_BLE/targets/TARGET_NORDIC/TARGET_NRF5/source/nRF5XCrypto.cpp:
```diff
@@ -0,0 +1,146 @@
+}
+
+bool LescCrypto::generate_shared_secret(
+    const ArrayView<const uint8_t>& peer_X,
```
I do agree that it is cheaper than a reference to an ArrayView however with a pointer to an uint8_t* it is easier to unintentionally pass an array of the wrong size. The ideal for this case would be a type that is a view to an array that encode the array size into the type.

pan- commented:
@paul-szczepanek-arm I think I've addressed your comments: 
- `ArrayView` can now encode the size of the array viewed in the type itself. Runtime encoding of the size is still supported as the default option.
- Debuging leftover has been removed.
- A small bit of documentation has been added to CryptoToolbox.

pan- commented:
@paul-szczepanek-arm Is this good enough to go in ?
